src/plugins/media/media.js

```diff
@@ -0,0 +1,261 @@
+ *    - data-media-autostop="true": Stop media (= pause and reset to start), when leaving its step.
+ *    - data-media-autopause="true": Pause media but keep current time when leaving its step.
+ *    
+ *    When these attributes are added to a step they are inherited by all media on this step. Of course 
+ *    this setting can be overwritten by adding different attributes to inidvidual media.
+ *    
+ *    The same rule applies when this attributes is added to the root element. Settings can be overwritten
+ *    for individual steps and media.
+ *
+ *    Examples:
+ *    - data-media-autostart="true" data-media-autostop="true": start media on enter, stop on leave, 
+ *      restart from beginning when re-entering the step.
+ *
+ *    - data-media-autostart="true" data-media-autopause="true": start media on enter, pause on leave, 
+ *      resume on re-enter
+ *
+ *    - data-media-autostart="true" data-media-autostop="true" data-media-autopause="true": start 
+ *      media on enter, stop on leave (stop overwrites pause).
+ *
+ *    - <div id="impress" data-media-autostart="true"> ... <div class="step" data-media-autostart="false">
+ *      All media is startet automatically on all steps except the one that has the data-media-autostart="false"
+ *      attribute.
+ *
+ *  - Pro tip: Use <audio onended="impress().next()"> or <video onended="impress().next()"> to proceed to the 
+ *    next step automatically, when the end of the media is reached.
+ *    
+ *
+ * Copyright 2018 Holger Teichert (@complanar)
+ * Released under the MIT license.
+ */
+/* global window, document */
+
+(function (document, window) {
+    "use strict";
+    var root, api, gc, gcAttributes;
+    
+    gcAttributes = [];
+    
+    // function names
+    var enhanceMediaNodes,
+        enhanceMedia,
+        removeMediaClasses,
+        onStepenter,
+        onStepleave,
+        onPlay,
+        onPause,
+        onEnded,
+        garbage;
+    
+    document.addEventListener("impress:init", function (event) {
+        root = event.target;
+        api = event.detail.api;
+        gc = api.lib.gc;
+        
+        if (root.dataset.mediaAutoplay === "" || root.dataset.mediaAutoplay === "true") {
+            root.mediaAutoplay = true;
+        } else {
+            root.mediaAutoplay = false;
+        }
+        if (root.dataset.mediaAutostop === "" || root.dataset.mediaAutostop === "true") {
+            root.mediaAutostop = true;
+        } else {
+            root.mediaAutostop = false;
+        }
+        if (root.dataset.mediaAutopause === "" || root.dataset.mediaAutopause === "true") {
+            root.mediaAutopause = true;
+        } else {
+            root.mediaAutopause = false;
+        }
+        
+        // this *must* be called only after setting the autopause, autoplay and autostop values
+        enhanceMedia();
+        
+        gc.pushCallback(garbage);
+    }, false);
+    
+    garbage = function () {
```

Comment:
gc/garbageCollector is not a perfect name, as it's not really garbage and not really a garbage collector.

To not perpetuate my poor choice of naming here, could you call this function something like `cleanup`, `cleanupAttributes` or `teardown` or `teardownCallback`.

Reply:
Good suggestion, thank you.